Refactor, bug fixes, fields

asweet-confluent opened the pull request:
So I forked `mhoffmann/main` (`009def4ae38ec30e5b40beeae26efe93484ab286`) back in June while working on a project that needed this grammar, and I ended up making quite a few changes.

- Added field names for convenience
- Fixed a few bugs. To be honest, I did most of this work over the summer, so I don't remember all the details of what I fixed, but I do recall fixing some issues related to heredocs as well as expressions being parsed incorrectly.
- General refactor of the grammar - IIRC, some things were changed just for my own convenience, others were changed to address bugs. It's very possible that these bugs have been fixed already in `main`.
- Lots of changes to the scanner. There were some bugs here, but I think I did this mainly to preserve whitespace in the tokens. I don't know if that's the "correct" approach to that problem, but it worked out for me.

I've been using this branch to create a Terraform interpreter with tree-sitter + Python, and the results are very good - 100% parity with Terraform in much of my internal corpus.

I don't expect you to merge these changes; this PR is more of a, "hey, look at this" (feel free to close it). Hopefully somebody gets some use out of it. Sorry that I don't remember all the details of what I changed. Hopefully the commit messages will give enough context, but feel free to drop a comment on anything that doesn't make sense and I'll try to remember what I was thinking.

MichaHoffmann commented:
Oh! very impressive - out of curiosity: what was your project?

I think I fixed the unary negation issue at some point 
and that might be the conflict! Can you maybe rebase on main and just clobber my change so that we can merge?

The only thing I fear is that this will break downstream uses since the tree changed so we might just need to release it as new major version.

MichaHoffmann commented:
Can you also update the Terraform dialect please? It's just a copy of the HCl parser / scanner where "HCl" is replaced by "Terraform" pretty much. That was because I was not able to use the HCl parser for the terraform language in neovim iirc.

MichaHoffmann commented:
> Done.

Sorry, comming back to this - we do have a couple test failures, can you look into them maybe?

asweet-confluent commented:
> Sorry, comming back to this - we do have a couple test failures, can you look into them maybe?

@MichaHoffmann  Done! (better late then never 😅)  I fixed a hang in the scanner and checked in a fuzzer test case for that (It's not particularly interesting, but it's probably good just to get the fuzzer  corpus plumbing set up anyway).

Since you said this would be a 2.0 release, I went through and cleaned up the grammar some more and tried to make things as consistent as possible. The PR description has been updated to reflect the changes.

Two open items regarding compatibility:
- `numeric_lit` [currently allows hex values ](https://github.com/tree-sitter-grammars/tree-sitter-hcl/blob/64ad62785d442eb4d45df3a1764962dafd5bc98b/make_grammar.js#L90)(in fact it allows even more than that, like `0xZZZZ`). There's a text fixture that exercises hex support [here](https://github.com/tree-sitter-grammars/tree-sitter-hcl/blob/64ad62785d442eb4d45df3a1764962dafd5bc98b/example/real_world_stuff/hashicorp/hashicorp%25vaultcommand%25server%25test-fixtures%25config_seals.hcl#L28). I suspect this fixture is older HCL1, [which does support `0x`/`0X` hex](https://github.com/hashicorp/hcl/blob/beb03eadfd3862ea77b8df8b5350c1f3d421f24c/hcl/scanner/scanner.go#L276-L287), while HCL2 [does not](https://github.com/hashicorp/hcl/blob/main/hclsyntax/spec.md#numeric-literals). It seems like the "right" thing would be a separate dialect for HCL1, but I'm tired and I'm not sure how much this matters.
- `config_file` is [currently ](https://github.com/tree-sitter-grammars/tree-sitter-hcl/blob/64ad62785d442eb4d45df3a1764962dafd5bc98b/make_grammar.js#L39-L40)`choice($.body, $.object)`. According to [the spec](https://github.com/hashicorp/hcl/blob/main/hclsyntax/spec.md#structural-elements), it should simply be `$.body`:
  ```ebnf
  ConfigFile   = Body;
  ```
  The reason we stray from the spec is to "also allow objects to handle .tfvars in json format". However, Terraform [doesn't](https://github.com/hashicorp/terraform/blob/d052145462e9b513211f335509bff31b86507127/internal/configs/parser.go#L80-L84) parse `.json` files with its HCL parser. A quick check reveals that HCL isn't a strict superset of JSON, e.g. the supported escape sequences are different:
  ```sh
  echo '{"x":"a\/b"}' | jq
  {
    "x": "a/b"
  }
  
  echo '{"x" = "a\/b"}' | terraform console
  ╷
  │ Error: Invalid escape sequence
  │
  │   on <console-input> line 1:
  │   (source code not available)
  │
  │ The symbol "/" is not a valid escape sequence selector.
  ╵
  ```
  I feel more strongly about this one:  I'd like to drop the `$.object` choice and the faux-JSON parsing support along with it. WDYT?
